### ISSUE TITLE
Fatal assert in V8 for Electron 17.0.0-alpha.4

### ISSUE THREAD

**Issue by daniel-koc** [electron/electron#32169]
The problem originates as change for https://bugs.chromium.org/p/chromium/issues/detail?id=1244145
The change fixes a problem introduced by https://chromium.googlesource.com/chromium/src/+/8a7275c6166e81b5285855afcbb85b38f219bca0
The actual code change that causes the problem is the factory_base.cc/h change from https://chromium-review.googlesource.com/c/v8/v8/+/3211575 , namely adding a NewArrayList() method call to FactoryBase<Impl>::NewScriptWithId(). The method is called with the default allocation type parameter of "kYoung", while the call ends up in the LocalFactory::AllocateRaw() method that only allows allocation in the "kOld" space:
 	v8_libbase.dll!v8::base::OS::Abort() Line 966	C++
 	v8_libbase.dll!V8_Fatal(const char * file, int line, const char * format, ...) Line 167	C++
 	v8_libbase.dll!v8::base::`anonymous namespace'::DefaultDcheckHandler(const char * file, int line, const char * message) Line 57	C++
>	v8.dll!v8::internal::LocalFactory::AllocateRaw(int size, v8::internal::AllocationType allocation, v8::internal::AllocationAlignment alignment) Line 43	C++
 	[Inline Frame] v8.dll!v8::internal::FactoryBase<v8::internal::LocalFactory>::AllocateRaw(int size, v8::internal::AllocationType allocation, v8::internal::AllocationAlignment alignment) Line 939	C++
 	v8.dll!v8::internal::FactoryBase<v8::internal::LocalFactory>::AllocateRawArray(int size, v8::internal::AllocationType allocation) Line 886	C++
 	[Inline Frame] v8.dll!v8::internal::FactoryBase<v8::internal::LocalFactory>::AllocateRawFixedArray(int length, v8::internal::AllocationType allocation) Line 903	C++
 	v8.dll!v8::internal::FactoryBase<v8::internal::LocalFactory>::NewFixedArrayWithFiller(v8::internal::Handle<v8::internal::Map> map, int length, v8::internal::Handle<v8::internal::Oddball> filler, v8::internal::AllocationType allocation) Line 128	C++
 	v8.dll!v8::internal::FactoryBase<v8::internal::LocalFactory>::NewFixedArray(int length, v8::internal::AllocationType allocation) Line 98	C++
 	v8.dll!v8::internal::FactoryBase<v8::internal::LocalFactory>::NewArrayList(int size, v8::internal::AllocationType allocation) Line 289	C++
 	v8.dll!v8::internal::FactoryBase<v8::internal::LocalFactory>::NewScriptWithId(v8::internal::Handle<v8::internal::PrimitiveHeapObject> source, int script_id) Line 252	C++
 	v8.dll!v8::internal::ParseInfo::CreateScript<v8::internal::LocalIsolate>(v8::internal::LocalIsolate * isolate, v8::internal::Handle<v8::internal::String> source, v8::internal::MaybeHandle<v8::internal::FixedArray> maybe_wrapped_arguments, v8::ScriptOriginOptions origin_options, v8::internal::NativesFlag natives) Line 235	C++
 	v8.dll!v8::internal::BackgroundCompileTask::Run() Line 1512	C++
 	blink_core.dll!blink::ScriptStreamer::RunScriptStreamingTask(std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>> task, blink::ScriptStreamer * streamer, blink::SourceStream * stream) Line 489	C++
 	blink_core.dll!base::internal::FunctorTraits<void (*)(std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>, blink::ScriptStreamer *, blink::SourceStream *),void>::Invoke<void (*)(std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>, blink::ScriptStreamer *, blink::SourceStream *),std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>,cppgc::internal::BasicCrossThreadPersistent<blink::ScriptStreamer,cppgc::internal::StrongCrossThreadPersistentPolicy,cppgc::internal::IgnoreLocationPolicy,cppgc::internal::DisabledCheckingPolicy>,blink::SourceStream *>(void(*)(std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>, blink::ScriptStreamer *, blink::SourceStream *) && function, std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>> && args, cppgc::internal::BasicCrossThreadPersistent<blink::ScriptStreamer,cppgc::internal::StrongCrossThreadPersistentPolicy,cppgc::internal::IgnoreLocationPolicy,cppgc::internal::DisabledCheckingPolicy> && args, blink::SourceStream * && args) Line 428	C++
 	blink_core.dll!base::internal::InvokeHelper<0,void>::MakeItSo<void (*)(std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>, blink::ScriptStreamer *, blink::SourceStream *),std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>,cppgc::internal::BasicCrossThreadPersistent<blink::ScriptStreamer,cppgc::internal::StrongCrossThreadPersistentPolicy,cppgc::internal::IgnoreLocationPolicy,cppgc::internal::DisabledCheckingPolicy>,blink::SourceStream *>(void(*)(std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>, blink::ScriptStreamer *, blink::SourceStream *) && functor, std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>> && args, cppgc::internal::BasicCrossThreadPersistent<blink::ScriptStreamer,cppgc::internal::StrongCrossThreadPersistentPolicy,cppgc::internal::IgnoreLocationPolicy,cppgc::internal::DisabledCheckingPolicy> && args, blink::SourceStream * && args) Line 708	C++
 	blink_core.dll!base::internal::Invoker<base::internal::BindState<void (*)(std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>, blink::ScriptStreamer *, blink::SourceStream *),std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>,cppgc::internal::BasicCrossThreadPersistent<blink::ScriptStreamer,cppgc::internal::StrongCrossThreadPersistentPolicy,cppgc::internal::IgnoreLocationPolicy,cppgc::internal::DisabledCheckingPolicy>,WTF::CrossThreadUnretainedWrapper<blink::SourceStream>>,void ()>::RunImpl<void (*)(std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>, blink::ScriptStreamer *, blink::SourceStream *),std::__1::tuple<std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>,cppgc::internal::BasicCrossThreadPersistent<blink::ScriptStreamer,cppgc::internal::StrongCrossThreadPersistentPolicy,cppgc::internal::IgnoreLocationPolicy,cppgc::internal::DisabledCheckingPolicy>,WTF::CrossThreadUnretainedWrapper<blink::SourceStream>>,0,1,2>(void(*)(std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>, blink::ScriptStreamer *, blink::SourceStream *) && functor, std::__1::tuple<std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>,cppgc::internal::BasicCrossThreadPersistent<blink::ScriptStreamer,cppgc::internal::StrongCrossThreadPersistentPolicy,cppgc::internal::IgnoreLocationPolicy,cppgc::internal::DisabledCheckingPolicy>,WTF::CrossThreadUnretainedWrapper<blink::SourceStream>> && bound, std::__1::integer_sequence<unsigned long long,0,1,2>) Line 781	C++
 	blink_core.dll!base::internal::Invoker<base::internal::BindState<void (*)(std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>, blink::ScriptStreamer *, blink::SourceStream *),std::__1::unique_ptr<v8::ScriptCompiler::ScriptStreamingTask,std::__1::default_delete<v8::ScriptCompiler::ScriptStreamingTask>>,cppgc::internal::BasicCrossThreadPersistent<blink::ScriptStreamer,cppgc::internal::StrongCrossThreadPersistentPolicy,cppgc::internal::IgnoreLocationPolicy,cppgc::internal::DisabledCheckingPolicy>,WTF::CrossThreadUnretainedWrapper<blink::SourceStream>>,void ()>::RunOnce(base::internal::BindStateBase * base) Line 750	C++
 	base.dll!base::OnceCallback<void ()>::Run() Line 143	C++
 	base.dll!base::TaskAnnotator::RunTaskImpl(base::PendingTask & pending_task) Line 137	C++
 	base.dll!base::TaskAnnotator::RunTask<`lambda at ../../base/task/thread_pool/task_tracker.cc:707:35'>(perfetto::StaticString event_name, base::PendingTask & pending_task, base::internal::TaskTracker::RunTaskImpl::<lambda_0> && args) Line 74	C++
 	base.dll!base::internal::TaskTracker::RunTaskImpl(base::internal::Task & task, const base::TaskTraits & traits, base::internal::TaskSource * task_source, const base::SequenceToken & token) Line 706	C++
 	base.dll!base::internal::TaskTracker::RunSkipOnShutdown(base::internal::Task & task, const base::TaskTraits & traits, base::internal::TaskSource * task_source, const base::SequenceToken & token) Line 691	C++
 	base.dll!base::internal::TaskTracker::RunTaskWithShutdownBehavior(base::internal::Task & task, const base::TaskTraits & traits, base::internal::TaskSource * task_source, const base::SequenceToken & token) Line 722	C++
 	base.dll!base::internal::TaskTracker::RunTask(base::internal::Task task, base::internal::TaskSource * task_source, const base::TaskTraits & traits) Line 552	C++
 	base.dll!base::internal::TaskTracker::RunAndPopNextTask(base::internal::RegisteredTaskSource task_source) Line 468	C++
 	base.dll!base::internal::WorkerThread::RunWorker() Line 379	C++
 	base.dll!base::internal::WorkerThread::RunPooledWorker() Line 267	C++
 	base.dll!base::internal::WorkerThread::ThreadMain() Line 247	C++
 	base.dll!base::`anonymous namespace'::ThreadFunc(void * params) Line 124	C++
 	[External Code]	
Electron has enabled the flag with intake to Chromium 98.0.4706.0: https://github.com/electron/electron/commit/bd10b19b0cdc46cdbadb570af89305e64541b679#diff-a7b57c23a681457fb05e511eb155bb37b431faa7e3c7523c6490ff79c3d46f27

**Comment by kromain** [comment]
I'm hitting the same assertion since updating to the 17-x-y branch on Windows, though not on my usual testing page. But loading almost any other "real" app or website triggers this. I wonder how this has not been noticed more widely yet.

**Comment by kromain** [comment]
I got it fixed by adding an allocation argument to `NewArrayList()` and passing `AllocationType::kOld` explicitly when called from `NewScriptWithId()`. That seems to do the trick for me, though I'm not sure what the implications of using `kOld` instead of `kYoung` really are. I assume it doesn't matter since it's a zero-sized array and a temporary workaround anyway, but I've uploaded the patch the owner of the original fix to comment on:

https://chromium-review.googlesource.com/c/v8/v8/+/3354087

Assuming it doesn't break things in an unexpected way, we can then backport the patch to the `main` and `17-x-y` branches here.
